graph/formats/rdf/rdf.go

```diff
@@ -0,0 +1,386 @@
+// be valid and include a scheme. The returned Term will not have
+// the UID set.
+func NewIRITerm(iri string) (Term, error) {
+	err := checkIRIText(iri)
+	if err != nil {
+		return Term{}, err
+	}
+	return Term{Value: escape("<", iri, ">")}, nil
+}
+
+// NewLiteralTerm returns a Term based on the literal text and an
+// optional qualifier which may either be a "@"-prefixed language
+// tag or a valid IRI. The text will be escaped if necessary and quoted,
+// and if an IRI is given it will be escaped if necessary. The returned
+// Term will not have the UID set.
+func NewLiteralTerm(text, qual string) (Term, error) {
+	text = escape(`"`, text, `"`)
+	if qual == "" {
+		return Term{Value: text}, nil
+	}
+	if strings.HasPrefix(qual, "@") {
+		err := checkLangText([]byte(qual))
+		if err != nil {
+			return Term{}, err
+		}
+		return Term{Value: text + qual}, nil
+	}
+	err := checkIRIText(qual)
+	if err != nil {
+		return Term{}, err
+	}
+	return Term{Value: text + escape("^^<", qual, ">")}, nil
+}
+
+func checkIRIText(iri string) error {
+	switch u, err := url.Parse(iri); {
+	case err != nil:
+		return err
+	case u.Scheme == "":
+		return fmt.Errorf("rdf: %w: relative IRI ref %q", ErrInvalidTerm, iri)
+	default:
+		return nil
+	}
+}
+
+// Parts returns the pars of the term and the kind of the term.
+// IRI node text is returned as a valid IRI with the quoting angle
+// brackets removed and escape sequences interpreted, and blank
+// nodes are stripped of the "_:" prefix.
+// When the term is a literal, qual will either be empty, an unescaped
+// IRI, or an RDF language tag prefixed with an @ symbol. The literal
+// text is returned unquoted and unescaped.
+func (t Term) Parts() (text, qual string, kind Kind, err error) {
+	return extract([]rune(t.Value))
+}
+
+// ID returns the value of the Term's UID field.
+func (t Term) ID() int64 { return t.UID }
+
+// Statement is an RDF statement. It implements the graph.Edge and graph.Line
+// interfaces.
+type Statement struct {
+	Subject   Term
+	Predicate Term
+	Object    Term
+	Label     Term
+}
+
+// String returns the RDF 1.1 N-Quad formatted statement.
+func (s *Statement) String() string {
+	if s.Label.Value == "" {
+		return fmt.Sprintf("%s %s %s .", s.Subject.Value, s.Predicate.Value, s.Object.Value)
+	}
+	return fmt.Sprintf("%s %s %s %s .", s.Subject.Value, s.Predicate.Value, s.Object.Value, s.Label.Value)
+}
+
+// From returns the subject of the statement.
+func (s *Statement) From() graph.Node { return s.Subject }
+
+// To returns the object of the statement.
+func (s *Statement) To() graph.Node { return s.Object }
+
+// ID returns the UID of the Predicate field.
+func (s *Statement) ID() int64 { return s.Predicate.UID }
+
+// ReversedEdge returns a graph.Edge with the subject and object of the
+// statement swapped. It is implemented to satisfy the graph.Edge
+// interface.
+func (s *Statement) ReversedEdge() graph.Edge {
+	r := *s
+	r.Subject, r.Object = r.Object, r.Subject
+	return &r
+}
```

Comment:
I'm having doubts about the behaviour of this and `ReversedLine`. ISTM given the absence of intrinsic meaning to the reversal of an RDF statement (without domain-specific knowledge) that this should return the receiver.

If the edge is not returned with the ends swapped, the documentation for `graph.Edge` should be altered so that the `ReversedEdge` method has:
```
    // ReversedEdge returns the edge reversal of the receiver
    // if a reversal is valid for the data type.
    // When a reversal is valid an edge of the same type as
    // the receiver with nodes of the receiver swapped should
    // be returned, otherwise the receiver should be returned
    // unaltered.
```

with the equivalent change for `graph.Line`.

Reply:
Having had some time to think about this, I think this is the correct thing to do. It has virtually no impact on tests within the graph packages (changing simple and multi edges and lines to return only the unreversed edge/line - to simulate a user using a graph that follows now-allowed semantics - causes failures in simple as expected since they test for this specific behaviour (not in multi - I should fix that), and in `graph.Copy` which makes sense given what the test is testing).

I've added an example that demonstrated how to make use of this feature.